Add delta-delta-p (ddp) tree inference approach

jroessler opened the pull request:
## Proposed changes

As suggested in #300  I implemented the delta-delta-p (in short: DDP) tree based inference method originally proposed in Hansotia and Rukstales: Incremental value modeling (2002). I simply added another evaluationFunction in _causalml/inference/tree/models.py_ called 'DDP'.

## Types of changes

- [ ]  Bugfix (non-breaking change which fixes an issue)
- [x]  New feature (non-breaking change which adds functionality)
- [ ]  Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ]  Documentation Update (if none of the other choices apply)

## Checklist

- [x]  I have read the [CONTRIBUTING](https://github.com/uber/causalml/blob/master/CONTRIBUTING.md) doc
- [x]  I have signed the [CLA](https://cla-assistant.io/uber/causalml)
- [x]  Lint and unit tests pass locally with my changes
- [ ]  I have added tests that prove my fix is effective or that my feature works
- [x]  I have added necessary documentation (if appropriate)
- [x]  Any dependent changes have been merged and published in downstream modules

## Further comments

There was no need to add an additional test for proving that the new tree based inference method works. You can simply use the _test_uplift_trees.py_ test and provide _evaluationFunction='DDP'_ when creating the _UpliftRandomForestClassifier_ and _UpliftTreeClassifier_ instances. I noticed that my test does not pass if _N_SAMPLE = 1000_ because the performance of random targeting was better than the DDP approach. However increasing _IN_SAMPLE_ to, for example, 10.000 worked.

If you need additional information, don't hesitate to ask me!

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/uber/causalml?pullRequest=327) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/uber/causalml?pullRequest=327) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/uber/causalml?pullRequest=327) it.</sub>

paullo0106 commented on causalml/inference/tree/models.py:
```diff
@@ -990,6 +1018,14 @@ def growDecisionTreeFrom(self, X, treatment, y, evaluationFunction, max_depth=10
+                        leftScore1 = evaluationFunction(leftNodeSummary, control_name=self.control_name)
+                        rightScore2 = evaluationFunction(rightNodeSummary, control_name=self.control_name)
+                        gain = np.abs(leftScore1 - rightScore2)
+                        gain_for_imp = (len(X_l) * leftScore1 - len(X_r) * rightScore2)
```
@jroessler I'm not familiar with this method, but just want to confirm if it's the case that `gain` takes the absolute difference while `gain_for_imp` doesn't

jroessler replied:
@paullo0106 Thanks for the pointer! Actually it's a mistake! It should be the absolute value for `gain_for_imp` as well. I'll fix it and resubmit.

jroessler commented:
I added raising an exception if the DDP approach is used with more than two treatment options as mentioned here:
https://github.com/uber/causalml/issues/300#issuecomment-831209598

Further, regarding the tests, I added a method called `generate_classification_data_two_treatments` which creates a synthetic data set with only two treatment options. You can use this method within `test_UpliftTreeClassifier` and `test_UpliftRandomForestClassifier` to verify that the DDP approach is working as excepted. _As a side note_: I realized that the results for all evaluation functions (not only for the DDP approach but also for KL, ED, and CHI) changed from test (RandomForestTest) to test. Thus, it seems that something with `random_state` is not working properly.

jeongyoonlee commented:
Thanks, @jroessler for the contribution.  It looks good to me. @t-tte, could you check this PR?

t-tte reviewed (approved):
Thanks for the contribution.

LGTM also.
